src/java.base/unix/classes/sun/nio/fs/UnixPath.java

```diff
@@ -857,13 +860,76 @@ public Path toRealPath(LinkOption... options) throws IOException {
             result = result.resolve(element);
         }
 
-        // check file exists (without following links)
+        // check whether file exists (without following links)
         try {
             UnixFileAttributes.get(result, false);
         } catch (UnixException x) {
             x.rethrowAsIOException(result);
         }
-        return result;
+
+        // Return if the file system is not both case insensitive and retentive
+        if (!fs.isCaseInsensitiveAndPreserving())
+            return result;
+
+        UnixPath path = fs.rootDirectory();
+
+        SecurityManager sm = System.getSecurityManager();
+
+        // Traverse the result obtained above from the root downward, leaving
+        // any '..' elements intact, and replacing other elements with the
+        // entry in the same directory which has an equal key
+        for (int i = 0; i < result.getNameCount(); i++ ) {
+            UnixPath element = result.getName(i);
+
+            // If the element is "..", append it directly and continue
+            if (element.toString().equals("..")) {
+                path = path.resolve(element);
+                continue;
+            }
+
+            // Derive full path to element and check readability
+            UnixPath elementPath = path.resolve(element);
+
+            // Derive element key
```

Comment:
You might want to re-visit the comments at L890 and 893 and there is no deviation, instead it just gets the file key. The API allows is to be null but the Unix implementation is always non-null.

Reply:
Revised in c193b3b48f56e2c581efaf212f2770c9add36a0a.